File: 1610-maximum-number-of-visible-points/1610-maximum-number-of-visible-points.js
/**
 * @param {number[][]} points
 * @param {number} angle
 * @param {number[]} location
 * @return {number}
 */
var visiblePoints = function(points, angle, location) {
  angle = angle * Math.PI / 180;
  const [x0, y0] = location;
  const angles = [];
  let [j, extra, ans] = [0, 0, 0];
  
  for (let [x, y] of points) {
    if (x == x0 && y == y0) {
      extra++;
      continue;
    }
    
    angles.push(Math.atan2(y - y0, x - x0));
  }
  
  angles
    .sort((a, b) => a - b)
    .forEach((x) => angles.push(x + Math.PI * 2));
  
  angles.forEach((x, i) => {
    while (x - angles[j] > angle) j++;
    
    ans = Math.max(ans, i - j + 1);
  })
  
  return ans + extra;
};

File: 1610-maximum-number-of-visible-points/1610-maximum-number-of-visible-points.py
class Solution:
    def visiblePoints(self, points: List[List[int]], angle: int, location: List[int]) -> int:
        angle = angle * math.pi / 180
        x0, y0 = location
        angles = []
        j = extra = count = 0
        for x, y in points:
            if x == x0 and y == y0:
                extra += 1
            else:
                angles.append(math.atan2(y - y0, x - x0))
        angles.sort()
        angles += [x + math.pi * 2 for x in angles]
        for i, x in enumerate(angles):
            while x - angles[j] > angle:
                j += 1
            count = max(count, i - j + 1)
        return count + extra